analyses/cms-open-data-ttbar/analysis.py

```diff
@@ -221,13 +221,21 @@ def book_histos(
     # Selecting events containing at least one lepton and four jets with pT > 25 GeV
     # Applying requirement at least one of them must be b-tagged jet (see details in the specification)
     df = (
-        df.Define("Electron_pt_mask", "Electron_pt>25")
-        .Define("Muon_pt_mask", "Muon_pt>25")
-        .Define("Jet_pt_mask", "Jet_pt>25")
-        .Filter("Sum(Electron_pt_mask) + Sum(Muon_pt_mask) == 1")
-        .Filter("Sum(Jet_pt_mask) >= 4")
+     df.Define(
+        "Electron_pt_mask", 
+        "Electron_pt > 30 && abs(Electron_eta) < 2.1 && Electron_sip3d < 4 && Electron_cutBased == 4")
+       .Define(
+        "Muon_pt_mask", 
+        "Muon_pt > 30 && abs(Muon_eta) < 2.1 && Muon_sip3d < 4 && Muon_tightId && Muon_pfRelIso04_all < 0.15")
+       .Define(
+        "Jet_pt_mask", 
+        "Jet_pt > 30 && abs(Jet_eta) < 2.4 && Jet_jetId == 6")
+       .Filter("Sum(Electron_pt_mask) + Sum(Muon_pt_mask) == 1")
+       .Filter("Sum(Jet_pt_mask) >= 4")
```

Comment:
nitpick: moving Define'd columns closer to point of usage:

```suggestion
     df.Define(
        "Electron_pt_mask", 
        "Electron_pt > 30 && abs(Electron_eta) < 2.1 && Electron_sip3d < 4 && Electron_cutBased == 4")
       .Define(
        "Muon_pt_mask", 
        "Muon_pt > 30 && abs(Muon_eta) < 2.1 && Muon_sip3d < 4 && Muon_tightId && Muon_pfRelIso04_all < 0.15")
       .Filter("Sum(Electron_pt_mask) + Sum(Muon_pt_mask) == 1")
       .Define(
        "Jet_pt_mask", 
        "Jet_pt > 30 && abs(Jet_eta) < 2.4 && Jet_jetId == 6")
       .Filter("Sum(Jet_pt_mask) >= 4")
```

Reply:
From my subjective point easier to read `Definition` block and then `Filtering` block

Reply:
I'd `Define` is better to keep separate from `Filter` because `Define` actually represents selecting "good" electrons, muons, and jets, i. e. particles, while `Filter` represents selecting "good" events. So they can be thought of as logically different parts.

Reply:
Another thing to consider is that calling `Define` before filter means potentially a cost in having to actually run the computation just to then discard the event right after.

Reply:
@vepadulano I don't think so? `Define`s are evaluated lazily and only if needed independently of where they appear.

Reply:
@andriiknu I suggested the change because the only place where `Electron_pt_mask` and `Muon_pt_mask` are used is in that one `Filter`.

So this is a cohesive piece of logic after which the reader can forget about these quantities:

```cpp
.Define("Electron_pt_mask", ..)
.Define("Muon_pt_mask", ..)
.Filter("Sum(Electron_pt_mask) + Sum(Muon_pt_mask) == 1")
```

I did mark this as a nitpick: I don't mind if you want to leave it as it is.

Reply:
actually now that I take a fresh look at this, I think this would be objectively clearer:

```cpp
.Define("n_good_electrons", "Sum(Electron_pt > 25)")
.Define("n_good_muons", "Sum(Muon_pt > 25)")
.Filter("n_good_electrons + n_good_muons == 1")
```

Reply:
...or maybe not :smile: up to you Andrii :)

Reply:
Yes, It's clearer
And one more thing: I don't think we need to keep names such as `Electron_pt_mask` or `Jet_pt_mask`, because now there is no any other mask in code. So, I'd used just  `Jet_mask` and `Jet_filetered` of `Jet_selected` to avoid complex names

Reply:
`Electron_pt_mask` and `Muon_pt_mask` already go away if you apply my last suggestion.

I like `Jet_mask` given there is no other masking of jets other than with the pt mask.

I don't know what `Jet_selected` or `Jet_masked` would refer to because we have different columns for each of the components `Jet_pt`, `Jet_eta`, etc.

Reply:
Ah, I realized. Yeah, it is correct we throw away electron and muon masks as separate colomns with your suggestions. But good muons and electrons itself will probably be needed for features. So, this good improvement for v1, but not sure about v2.
About Jet_selected I meant first of all Jet_pt_selected and Jet_btagCSVV_selected

Reply:
feel free to rearrange things as you prefer, I trust you -- and this PR is already approved so you can merge it when you are satisfied :smile:

Reply:
Okay, I already rearranged these things in the next commits, but these are coming in the next PR that should include ML part. Thank you